### PR TITLE
[ENH] Introduce `BaseFoundationForecaster` to unify zero-shot foundation models

### LINKED ISSUE #10498: [ENH] Unify pretrained/foundation forecasters under a shared base class
Detailed Plan: https://hackmd.io/@geetu/B1Ccj-HXGx

`sktime` now has many pretrained/foundation-model forecasters, including Chronos, TimesFM, TinyTimeMixer, LagLlama, MOIRAI, Timer, Toto, Kronos, WindFM, and generic Hugging Face Transformers models.

These wrappers currently duplicate a lot of infrastructure: model loading, caching, device handling, serialization, prediction formatting, quantile handling, and fine-tuning/pretraining setup. Their APIs also expose device, PEFT, quantization, and training behavior inconsistently.

This issue proposes introducing a shared internal base class for foundation-model forecasters.

**Proposal**

Add an internal `BaseFoundationForecaster` that subclasses `BaseForecaster` and owns the common lifecycle:

- `_fit`, `_pretrain`, `_predict`, `_predict_quantiles`, `_predict_interval`
- model loading and cache lookup
- serialization/reload behavior
- device/dtype/quantization config handling
- normalized forecast result formatting

Concrete estimators should subclass it directly and implement only model-specific hooks, such as:

```python
class BaseFoundationForecaster(BaseForecaster):
    def _get_model_load_config(self): ...
    def _load_model(self, load_config, tune_config): ...
    def _make_context(self, y, X, fh, cutoff, handle): ...
    def _predict_native(self, handle, context, request): ...
```

Supporting internal objects could include:

`ModelLoadConfig`, `FineTuneConfig`, `ParameterEfficientTuneConfig`, `InferenceConfig`, `FoundationModelSpec`, `ModelHandle`, `ModelContext`, `ForecastRequest`, `ForecastResult`

**Acceptance Criteria**

- A shared internal `BaseFoundationForecaster` exists.
- Common load/cache/device/serialization/prediction-formatting logic is implemented once.
- At least 2-3 representative foundation forecasters are migrated.
- Existing public estimator constructors remain backward compatible.
- Quantization, PEFT, pretraining, zero-shot inference, and probabilistic outputs have normalized internal config/result objects.
- Added test coverage.
- Migrated estimators keep their existing user-facing behavior.

**Summary**

This would centralize duplicated infrastructure across foundation-model wrappers, especially loading, caching, device handling, serialization, prediction formatting, quantization, PEFT, and pretraining.

pros:
- a lot of bloated code gets removed
- expanding capabilities will be easier
- interfacing new models will be easier
- end-to-end testing of each component can be ensured in a better way given we have a base model contract atleast
- improving inference/speed-ups and training will be easy at a bigger scale, doesn't need to be impelmentated for each estimator

cons:
- hard design and implementation problem
- not sure if the output is going to be very clean, how generalized the models can be
- main cost is that this is a difficult abstraction: current models are heterogeneous, including Hugging Face models, pipelines, GluonTS/Lightning models, JAX models, tokenizer/model pairs, and domain-specific forecasters

The refactor should be incremental, starting with a few representative models before migrating the full ecosystem.

### PR DESCRIPTION
**Pull request by geetu040**
Fixes #10498

### Summary

This PR introduces `BaseFoundationForecaster`, a shared implementation for zero-shot foundation-model adapters.

Previously, every adapter independently handled model loading, caching, context storage, forecasting horizons, inference setup, serialization, and output formatting. These responsibilities are now centralized, leaving child classes responsible primarily for model-specific loading and inference.

### Design

The new foundation infrastructure consists of:

- `FoundationModelSpec`: separates model identity, loading options, and prediction options.
- `ModelHandle`: contains reusable native models, tokenizers, or pipelines.
- `ForecastResult`: provides a common representation for mean, median, and quantile forecasts.
- `BaseFoundationForecaster`: manages the shared fit and prediction lifecycle.
- A process-local model cache and common pandas formatting utilities.

Child adapters normally implement only:

- `_load_model`: load reusable backend objects.
- `_inference`: convert pandas context to native input and return `ForecastResult`.

Optional hooks support fit-derived configuration, custom cache keys, and model-specific output behavior.

A child estimator would look something like this:
```python
class MyForecaster(BaseFoundationForecaster):
    def __init__(self, model_path, device="auto"):
        self.model_path = model_path
        self.device = device

        model_spec = FoundationModelSpec(
            model_path=model_path,
            device=device,
            load_extra_kwargs={...},
            predict_extra_kwargs={...},
        )
        super().__init__(model_spec=model_spec)

	def _load_model(self):
		model = NativeModel.from_pretrained(
			self.model_spec.model_path,
			**self.model_spec.load_extra_kwargs,
		)
		return ModelHandle(model=model)

	def _inference(
		self,
		handle,
		context_y,
		context_X,
		future_X,
		pred_len,
		fh,
		alpha=None,
	):
		predictions = handle.model.predict(
			context_y.to_numpy(),
			prediction_length=pred_len,
			**self.model_spec.predict_extra_kwargs,
		)
		return ForecastResult(mean=predictions)
```

### Benefits

- Removes duplicated infrastructure from model adapters.
- Makes new foundation models easier to implement and review.
- Provides consistent horizon, point, and quantile formatting.
- Safely reuses expensive model objects between compatible estimators.
- Keeps fitted-series state separate from shared model state.
- Supports reproducible Torch inference and lazy model reload after serialization.
- Makes cache identity explicit by separating loading and prediction parameters.

### Affected forecasters

The current branch migrates 16 forecasters:

- Aurora
- Chronos
- Chronos-2
- CiscoTSM
- Falcon-TST
- FlowState
- Kronos
- MIRA
- Time-LLM
- TimeMoE
- Timer
- Timer-S1
- TimesFM
- TiRex
- Toto
- WindFM

### CODE REVIEW

File: sktime/forecasting/timesfm.py [hunk 1]
```diff
@@ -224,17 +226,35 @@ def __init__(
...
 
         This method should be used for setting dynamic tags only.
         """
-        if not self.ignore_deps:
+        super().__dynamic_tags__()
+        if not self.model_spec.ignore_deps:
             if self.use_source_package:
                 # Use timesfm with a version bound if use_source_package is True
-                # todo 1.2.0: Regularly check whether timesfm version can be updated
+                # todo 1.1.0: Regularly check whether timesfm version can be updated
```
Comment (fkiraly) [inline sktime/forecasting/timesfm.py:257]: accidental revert. Please check for more accidental reverts.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks - overall this is an ok uniformization.

Though I suspect this is quite convoluted design-by-AI.

The main issue is that this is unnecessarily complicated, introduces coupling, and is hard to extend. I fear if we merge this, no one will be able to add a foundation model anymore.

Examples of things that are anti-patterns or unnecessarily convoluted:
* there is no clear or simple extender API according to the template pattern. It is not clearly documented what method a person has to add.
* I think you have an anti-pattern throughout where you force inputs and returns of methods to be classes or dicts. This is an anti-pattern since (a) it feels unnecessary, and (b) an extender has to remember which class they are receiving or have to wrap outputs in. Examples:
    * the implementer has to return a `ModelHandle` from `_load_model`. This feels unnecessary, because the boilerplate reading `_load_model` could simply wrap it in `ModelHandle` and the extender then does not need to deal with it. However, with this pattern, the extender has to import it, and wrap stuff in it, without that having any additional effect for an extender.
    * similarly, the implementer has to call `super._init__` with a `FoundationModelSpec`. Again, this feels like an unnecssary chore.
    * same with `ForecastResult`, I gather you see the anti-pattern now. The class feels entirely unnecessary and could be handled by `format_quantile_result` etc.
    * kwargs are wrapped in `predict_extra_kwargs`, but then they have to be manually unwrapped later on, again variable by variable. It is unclear why this dict is populated in the first place, instead of just accessing `self.paramname` where it would be stored in `__init__` under the usual contract. The extender has to guess that `model_spec` is saved to self, then access some specific parameters from `self.model_spec.predict_extra_kwargs["paramname"]` instead of the more obvious `self.paramname`.
* I think the multiton pattern is more intuitive than the `ModelHandle` that is never explained properly. Either way, why duplicate the caching function instead of extend it with hashing?

I think there are also design problems:
* how would this work with zero-shot models that have a `_pretrain` capability, or a proba capability? This does not look clear to me.
* `_inference` looks like an amalgamation of `_predict` and `_predict_quantiles` to me, with the difference that the extender now has to wrap everything in a `ForecastResult`.
* `_foundation_predict` looks unnecessary, because it does two things: it fuses the multiton pattern into the base class, introducing more coupling; and it normalizes `alpha`, which `_predict_quantiles` already does afaik.

**Comment by geetu040** [comment]
@fkiraly Thanks for the review. Here are my thoughts.

---

> Thanks - overall this is an ok uniformization.
>
> Though I suspect this is quite convoluted design-by-AI.
>
> The main issue is that this is unnecessarily complicated, introduces coupling, and is hard to extend. I fear if we merge this, no one will be able to add a foundation model anymore.

I think we did agree on this design. I presented the [hackmd](https://hackmd.io/MHyFtJuRRUODy4FA0x8yig) in meetings a few times as it went through iterations, and I only started implementing after you gave the green light.

That said, I'm fine with redesigning it. This is the first time it's really been reviewed in this much detail, and better now than any later.

---

> * there is no clear or simple extender API according to the template pattern. It is not clearly documented what method a person has to add.

Agreed, this should be better. It's close to the `BaseDeepNetworkPyTorch` pattern, which we don't have an extension template for either, but I'll make the docstrings in `BaseFoundationForecaster` a lot more explicit about which methods you need to implement and what they should return.

---

> * I think you have an anti-pattern throughout where you force inputs and returns of methods to be classes or dicts. [...]
>   * the implementer has to return a `ModelHandle` from `_load_model`. [...]
>   * same with `ForecastResult`, I gather you see the anti-pattern now. [...]

So the main problem is the dataclasses we force the inputs/outputs into. I think removing them fixes most of the points here. My thinking was that adding a new model would be cleaner if you only had to deal with a few well documented dataclasses, but if it ends up confusing and hard to follow, it's not doing its job and I'll remove it.

---

>   * similarly, the implementer has to call `super._init__` with a `FoundationModelSpec`. Again, this feels like an unnecssary chore.
>   * kwargs are wrapped in `predict_extra_kwargs`, but then they have to be manually unwrapped later on [...]

I don't love this pattern either, but I thought it would pay off long term.

The problem is that arguments aren't uniform across the foundation models, even when they conceptually do the same thing. We wrap third-party packages, so the terminology differs a lot, and we never had a common pattern for implementing zero-shot foundation models. So we ended up with estimators that do the same thing but behave differently. Device naming/setting is one example (#11166), and there are a bunch of other inconsistencies listed in #10498.

This was an attempt to unify the interface and bring at least some uniformity to the internal implementations. That's why init args and predict args are separated: one group is for loading/caching, the other for inference.

If we keep this structure, then later on, as we go through deprecations and move away from the third-party implementations, the grouping becomes useful. Each model shrinks to a few lines, the shared logic lives in `BaseFoundationForecaster`, and once load/infer are guaranteed to be uniform we can build things like PEFT tuning and quantization on top without changing each estimator.

---

> * I think the multiton pattern is more intuitive than the `ModelHandle` that is never explained properly. Either way, why duplicate the caching function instead of extend it with hashing?

Agreed. I didn't have a strong reason to move away from the multiton pattern, or at least I can't remember one now. I'll try going back to it and extending it with hashing instead of duplicating the caching.

---

> * how would this work with zero-shot models that have a `_pretrain` capability, or a proba capability? This does not look clear to me.

You can still use this as the base class for those models and nothing breaks. `_pretrain` works independently of this.

For proba: the custom `_predict_proba` we had after #10566 did show some design flaws, which is why I ended up changing the alpha handling you mention below. If `_predict_proba` just uses `_predict_quantiles` internally, which is what most forecasters do right now, it should be fine. But I agree this part of the design is a limitation and needs a better way forward. I'd actually prefer proba to also go through `_foundation_predict`, like `_predict` and `_predict_quantiles` do. Reasoning below.

---

> * `_inference` looks like an amalgamation of `_predict` and `_predict_quantiles` to me, with the difference that the extender now has to wrap everything in a `ForecastResult`.

That's the idea, yes. Foundation models don't run separate forward passes for point and quantile forecasts. I think we can safely treat it as a rule that they do a single forward pass, and the output is just post-processed differently for point vs quantile. So I kept the shared part in one function. Random seed, device placement, eval mode and torch compiled graphs should all live in that common inference function too.

---

> * `_foundation_predict` looks unnecessary, because it does two things: it fuses the multiton pattern into the base class, introducing more coupling; and it normalizes `alpha`, which `_predict_quantiles` already does afaik.

Yes, it basically wraps inference: `_inference` does the actual torch-level forward pass, and `_foundation_predict` handles the sktime side. We could merge them, I just think it might end up less clean. But I don't strongly disagree.

---

This also aligns with our discussion about handling soft-deps, moving away from third-party implementations, so let's discuss this in the tech or community meetings to guess the future of this work.
